compiler/luci/pass/src/QuantizeWeightsWithGPTQPass.cpp

```diff
@@ -0,0 +1,674 @@
+    {
+      y[i] -= L[i * num_size + j] * y[j];
+    }
+    y[i] /= L[i * num_size + i];
+  }
+}
+
+void backward_substitution(const std::vector<float> &U, const std::vector<float> &y,
+                           std::vector<float> &x, int num_size)
+{
+  for (int i = num_size - 1; i >= 0; --i)
+  {
+    x[i] = y[i];
+    for (int j = i + 1; j < num_size; ++j)
+    {
+      x[i] -= U[i * num_size + j] * x[j];
+    }
+    x[i] /= U[i * num_size + i];
+  }
+}
+
+void cholesky_inverse(std::vector<float> &L, uint32_t num_size)
+{
+  std::vector<float> L_inv(L.size());
+  std::vector<float> H_inv(L.size());
+
+  std::vector<float> e(num_size, 0);
+  std::vector<float> col(num_size, 0);
+  std::vector<float> temp(num_size, 0);
+
+  for (uint32_t i = 0; i < num_size; ++i)
+  {
+    fill(e.begin(), e.end(), 0.0);
+    e[i] = 1.0;
+
+    forward_substitution(L, e, temp, num_size);
+
+    for (uint32_t j = 0; j < num_size; ++j)
+    {
+      L_inv[j * num_size + i] = temp[j];
+    }
+  }
+
+  for (uint32_t i = 0; i < num_size; i++)
+  {
+    for (uint32_t j = 0; j < i; j++)
+    {
+      float tmp = L[i * num_size + j];
+      L[i * num_size + j] = L[j * num_size + i];
+      L[j * num_size + i] = tmp;
+    }
+  }
+
+  for (uint32_t i = 0; i < num_size; ++i)
+  {
+    fill(e.begin(), e.end(), 0.0);
+    fill(col.begin(), col.end(), 0.0);
+    e[i] = 1.0;
+    for (uint32_t j = 0; j < num_size; j++)
+    {
+      col[j] = L_inv[j * num_size + i];
+    }
+    backward_substitution(L, col, temp, num_size);
+    for (uint32_t j = 0; j < num_size; ++j)
+    {
+      H_inv[j * num_size + i] = temp[j];
+    }
+  }
+  for (uint32_t i = 0; i < L.size(); i++)
+  {
+    L[i] = H_inv[i];
+  }
+}
+
+void cal_minmax_per_channel(CircleConst *node, std::vector<float> &min, std::vector<float> &max)
+{
+  loco::TensorShape dimension;
+  dimension.rank(4);
+  int32_t channel_dim_index{0};
+
+  if (!get_channel_dim_index(node, dimension, channel_dim_index))
+  {
+    assert(false);
+    return;
+  }
+  auto size = dimension.dim(channel_dim_index).value();
+
+  std::vector<bool> has_min_max_value(size, false);
+  min.resize(size);
+  max.resize(size);
+
+  auto cal_minmax = [&](uint32_t *indices, loco::TensorShape &dimension, int channel_dim_index) {
+    int channel_idx = indices[channel_dim_index];
+    auto data = node->at<loco::DataType::FLOAT32>(cal_offset(dimension, indices));
+    if (has_min_max_value[channel_idx])
+    {
+      min[channel_idx] = data < min[channel_idx] ? data : min[channel_idx];
+      max[channel_idx] = data > max[channel_idx] ? data : max[channel_idx];
+    }
+    else
+    {
+      min[channel_idx] = data;
+      max[channel_idx] = data;
+      has_min_max_value[channel_idx] = true;
+    }
+  };
+
+  iterate_per_channel_with_order(node, cal_minmax, false);
+}
+
+void compute_asym_scale_zp(float min, float max, float &scaling_factor, int64_t &zp,
+                           float &nudged_min, float &nudged_max, int32_t k_max_scale)
```

Comment:
looks like some arguments are input and some are outputs.
it would be nice to add some comments about this.
`k_max_scale` would be better to group with other inputs.